### PR TITLE
adding parity_netPeers

### PR DESCRIPTION
**Pull request by marcindsobczak**
# Implemented:
* JsonRpc response
* CLI response
* returning number of active peers
* returning number of connected peers
* returning max number of active peers
* returning PeerInfo[] object (array of all active peers)
	* returning peer id
	* returning peer name
	* returning list of capabilities
	* returning PeerNetworkInfo object
		* returning peer local address
		* returning peer remote address
	* returning EthProtocolInfo
		* returning protocol version
		* returning difficulty
		* returning head (headhash)
* test for standard case, Node=null, Session=null

### CODE REVIEW

File: src/Nethermind/Nethermind.Cli/Modules/ParityCliModule.cs [hunk 1]
```diff
@@ -43,5 +43,8 @@ public ParityCliModule(ICliEngine engine, INodeManager nodeManager) : base(engin
         
         [CliFunction("parity", "setEngineSignerSecret", Description = "Sets an authority account for signing consensus messages.")]
         public bool SetEngineSignerSecret(string privateKey) => NodeManager.Post<bool>("parity_setEngineSignerSecret", privateKey).Result;
+        
+        [CliFunction("parity", "netPeers", Description = "Returns connected peers. Peers with non-empty protocols have completed handshake.")]
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Cli/Modules/ParityCliModule.cs:47]: As its parameterless we can make it CliProperty

File: src/Nethermind/Nethermind.JsonRpc/Modules/Parity/ParityNetPeers.cs [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+        public PeerInfo[] Peers { get; set; }
+        
+        public ParityNetPeers(IPeerManager peerManager)
+        {
+            Active = peerManager.ActivePeers.Count;
+            Connected = peerManager.ConnectedPeers.Count;
+            Max = peerManager.MaxActivePeers;
+            Peers = peerManager.ActivePeers.Select(p => new PeerInfo(p)).ToArray();
+        }
+    }
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.JsonRpc/Modules/Parity/ParityNetPeers.cs:37]: ParityNetPeers shouldn't rely on IPeerManager, maybe abstract it to some builder method

File: src/Nethermind/Nethermind.JsonRpc/Modules/Parity/PeerNetworkInfo.cs [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+        public string RemoteAddress { get; set; }
+
+        public PeerNetworkInfo(Peer peer)
+        {
+            LocalAddress = peer.Node.Host;
+            RemoteAddress = peer.InSession != null ? peer.InSession.State != SessionState.New
+                    ? peer.InSession.RemoteHost : "Handshake"
+                : peer.OutSession != null ? peer.OutSession.State != SessionState.New
+                    ? peer.OutSession.RemoteHost : "Handshake"
+                : null;
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.JsonRpc/Modules/Parity/PeerNetworkInfo.cs:35]: This is hard to read. Please add method per session as you duplicate logic

File: src/Nethermind/Nethermind.JsonRpc/Modules/Parity/PeerInfo.cs [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+        public string Name { get; set; }
+        public List<Capability> Caps { get; set; }
+        public PeerNetworkInfo Network { get; set; }
+        public string Protocols { get; set; }
+        
+        public PeerInfo(Peer peer)
+        {
+            if (peer.Node == null)
+            {
+                throw new ArgumentException(
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.JsonRpc/Modules/Parity/PeerInfo.cs:37]: Do we want to fail here? If it was in some builder method we could just skip it?

File: src/Nethermind/Nethermind.Network/PeerManager.cs [hunk 5]
```diff
@@ -96,8 +96,9 @@ public class PeerManager : IPeerManager
 
         public IReadOnlyCollection<Peer> ActivePeers => _activePeers.Values.ToList().AsReadOnly();
         public IReadOnlyCollection<Peer> CandidatePeers => _peerPool.CandidatePeers.ToList();
+        public IReadOnlyCollection<Peer> ConnectedPeers => _activePeers.Values.Where(IsConnected).ToList().AsReadOnly();
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Network/PeerManager.cs:99]: Maybe just IEnumerable? Not sure myself.

File: src/Nethermind/Nethermind.JsonRpc/Modules/Parity/PeerInfo.cs [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+namespace Nethermind.JsonRpc.Modules.Parity
+{
+    public class PeerInfo
+    {
+        public string Id { get; set; }
+        public string Name { get; set; }
+        public List<Capability> Caps { get; set; }
+        public PeerNetworkInfo Network { get; set; }
+        public string Protocols { get; set; }
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.JsonRpc/Modules/Parity/PeerInfo.cs:31]: Caps and protocols are not populated yet

**Review by LukaszRozmej — CHANGES_REQUESTED** [review]
Caps and protocols are not populated yet.
No tests yet.
Overall looks good.

File: src/Nethermind/Nethermind.JsonRpc/Modules/Parity/ParityNetPeers.cs (hunk 2 above)
Comment (tkstanczak) [reply src/Nethermind/Nethermind.JsonRpc/Modules/Parity/ParityNetPeers.cs:37]: Why not @LukaszRozmej ?

File: src/Nethermind/Nethermind.Network/PeerManager.cs (hunk 5 above)
Comment (tkstanczak) [reply src/Nethermind/Nethermind.Network/PeerManager.cs:99]: better follow the rules as in others, IEnumerable would be dangerous here because we have multiple threads operating on some of those lists

File: src/Nethermind/Nethermind.JsonRpc/Modules/Parity/PeerInfo.cs (hunk 6 above)
Comment (tkstanczak) [reply src/Nethermind/Nethermind.JsonRpc/Modules/Parity/PeerInfo.cs:31]: Caps are protocols, then what are protocols here?

File: src/Nethermind/Nethermind.JsonRpc/Modules/Parity/ParityNetPeers.cs (hunk 2 above)
Comment (tkstanczak) [reply src/Nethermind/Nethermind.JsonRpc/Modules/Parity/ParityNetPeers.cs:37]: I see

Comment (tkstanczak) [reply src/Nethermind/Nethermind.JsonRpc/Modules/Parity/ParityNetPeers.cs:37]: Yeah, this should not be in the DTO

File: src/Nethermind/Nethermind.JsonRpc/Modules/Parity/PeerNetworkInfo.cs (hunk 3 above)
Comment (marcindsobczak) [reply src/Nethermind/Nethermind.JsonRpc/Modules/Parity/PeerNetworkInfo.cs:35]: Actually State > New condition changes nothing, as
0 - New
1 - HandshakeComplete
2 - Initialized
3 - DisconnectingProtocols
4 - Disconnecting
5 - Disconnected
Yesterday we wondered about it and established, that "Handshake" should be displayed for New, and RemoteHost for not New.
When written current way it's not possible to display "Handshake" to state other than New, if the condition is fulfilled (State != New), it displays RemoteHost, no matter if state is 1, 2, 3, 4 or 5, same would be with condition State > New.
I believe it should works like that, but correct me if for disconnecting/disconnected (3, 4, 5) it should display something different than RemoteHost. And for sure I will rewrite it more read-friendly way.

File: src/Nethermind/Nethermind.JsonRpc/Modules/Parity/PeerInfo.cs (hunk 6 above)
Comment (marcindsobczak) [reply src/Nethermind/Nethermind.JsonRpc/Modules/Parity/PeerInfo.cs:31]: Caps should display list of capabilities (trying hard to add it but failing as far).
Protocols should return PeerProtocolsInfo object, where is one more object EthProtocolInfo. And inside the second one should be:
-ethereum protocol version,
-difficulty (peer total difficulty if known),
-head (SHA3 of peer best block hash)

**Comment by LukaszRozmej** [comment]
For now this should be PR Draft.

**Review by tkstanczak — CHANGES_REQUESTED** [review]
tests missing

File: src/Nethermind/Nethermind.JsonRpc/Modules/Parity/EthProtocolInfo.cs [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+    public class EthProtocolInfo
+    {
+        [JsonProperty("version", Order = 0)]
+        public byte Version { get; set; }
+        
+        [JsonProperty("difficulty", Order = 1)]
+        public UInt256 Difficulty { get; set; }
+        
+        [JsonProperty("head", Order = 2)]
+        public string HeadHash { get; set; }
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.JsonRpc/Modules/Parity/EthProtocolInfo.cs:33]: Keccak, not string?

File: src/Nethermind/Nethermind.JsonRpc/Modules/Parity/EthProtocolInfo.cs [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+
+        public EthProtocolInfo(Peer peer)
+        {
+            Version = peer.InSession?.Node.EthProtocolVersion ?? peer.OutSession?.Node.EthProtocolVersion ?? 0;
+            Difficulty = peer.InSession?.Node.Difficulty ?? peer.OutSession?.Node.Difficulty ?? 0;
+            HeadHash = peer.InSession?.Node.HeadHash?.ToString() ?? peer.OutSession?.Node.HeadHash?.ToString();
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.JsonRpc/Modules/Parity/EthProtocolInfo.cs:39]: just use Keccak here, EthJson can convert

File: src/Nethermind/Nethermind.JsonRpc/Modules/Parity/PeerInfo.cs [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+using Nethermind.Network;
+using Nethermind.Stats.Model;
+using Newtonsoft.Json;
+
+namespace Nethermind.JsonRpc.Modules.Parity
+{
+    public class PeerInfo
+    {
+        [JsonProperty("id", Order = 0)]
+        public string Id { get; set; }
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.JsonRpc/Modules/Parity/PeerInfo.cs:28]: would it work as a PublicKey?

File: src/Nethermind/Nethermind.JsonRpc/Modules/Parity/PeerInfo.cs [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+        [JsonProperty("network", Order = 3)]
+        public PeerNetworkInfo Network { get; set; }
+        
+        [JsonProperty("protocols", Order = 4)]
+        public Dictionary<string, EthProtocolInfo> Protocols { get; set; }
+        
+        public PeerInfo(Peer peer)
+        {
+            Id = peer.InSession?.RemoteNodeId.ToString() ?? peer.OutSession?.RemoteNodeId.ToString();
+            Name = peer.Node.ClientId;
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.JsonRpc/Modules/Parity/PeerInfo.cs:45]: what if Node is null?

File: src/Nethermind/Nethermind.JsonRpc/Modules/Parity/PeerInfo.cs [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+        public List<Capability> Caps { get; set; }
+        
+        [JsonProperty("network", Order = 3)]
+        public PeerNetworkInfo Network { get; set; }
+        
+        [JsonProperty("protocols", Order = 4)]
+        public Dictionary<string, EthProtocolInfo> Protocols { get; set; }
+        
+        public PeerInfo(Peer peer)
+        {
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.JsonRpc/Modules/Parity/PeerInfo.cs:43]: better store session as a local variable, otherwise the session may change mid way (multithreaded) and then you retrieve a different Session object each time

File: src/Nethermind/Nethermind.JsonRpc/Modules/Parity/PeerNetworkInfo.cs [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+    public class PeerNetworkInfo
+    {
+        [JsonProperty("localAddress", Order = 0)]
+        public string LocalAddress { get; set; }
+        
+        [JsonProperty("remoteAddress", Order = 1)]
+        public string RemoteAddress { get; set; }
+
+        public PeerNetworkInfo(Peer peer)
+        {
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.JsonRpc/Modules/Parity/PeerNetworkInfo.cs:33]: store session as a local variable first (InSession ?? OutSession)

File: src/Nethermind/Nethermind.JsonRpc/Modules/Parity/PeerInfo.cs (hunk 10 above)
Comment (tkstanczak) [reply src/Nethermind/Nethermind.JsonRpc/Modules/Parity/PeerInfo.cs:47]: session?

Comment (tkstanczak) [reply src/Nethermind/Nethermind.JsonRpc/Modules/Parity/PeerInfo.cs:47]: I mean ? would check for null

**Review by tkstanczak — CHANGES_REQUESTED** [review]
we need tests for this

File: src/Nethermind/Nethermind.JsonRpc/Modules/Parity/ParityModule.cs [hunk 13]
```diff
@@ -40,6 +41,8 @@ public class ParityModule : IParityModule
         private readonly IEnode _enode;
         private readonly ISignerStore _signerStore;
         private readonly IKeyStore _keyStore;
+        private readonly IPeerManager _peerManager;
+        private ParityNetPeers _parityNetPeers;
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.JsonRpc/Modules/Parity/ParityModule.cs:45]: _parityNetPeers shoudln't be a field in class. It can be a variable in parity_netPeers method.

**Review by tkstanczak — CHANGES_REQUESTED** [review]
almost ready but we need a few more tests that will cover capabilities as well as the edge cases

File: src/Nethermind/Nethermind.JsonRpc/Modules/Parity/PeerInfo.cs [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+            ISession session = peer.InSession ?? peer.OutSession;
+            PeerNetworkInfo peerNetworkInfo = new PeerNetworkInfo();
+            EthProtocolInfo ethProtocolInfo = new EthProtocolInfo();
+            List<Capability> capabilities = new List<Capability>();
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.JsonRpc/Modules/Parity/PeerInfo.cs:49]: we can drop capabilities variable

File: src/Nethermind/Nethermind.JsonRpc/Modules/Parity/PeerInfo.cs [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+                        ethProtocolInfo.Difficulty = syncPeer.TotalDifficulty;
+                        ethProtocolInfo.HeadHash = syncPeer.HeadHash;
+                    }
+                }
+                
+                if (session.TryGetProtocolHandler(Protocol.P2P, out var p2PHandler))
+                {
+                    if (p2PHandler is P2PProtocolHandler p2PProtocolHandler)
+                    {
+                        capabilities = p2PProtocolHandler.AgreedCapabilities;
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.JsonRpc/Modules/Parity/PeerInfo.cs:77]: we  can fill caps here directly

**Review by LukaszRozmej — APPROVED** [review]
Few minor changes mainly of variable usage, but other than that its great.

File: src/Nethermind/Nethermind.JsonRpc/Modules/Parity/PeerInfo.cs [hunk 16]
```diff
@@ -0,0 +1,93 @@
...
+                    }
+                }
+            }
+
+            foreach (Capability capability in capabilities)
+            {
+                Caps.Add(capability.ProtocolCode + "/" + capability.Version);
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.JsonRpc/Modules/Parity/PeerInfo.cs:84]: string.Concat

File: src/Nethermind/Nethermind.JsonRpc/Modules/Parity/PeerInfo.cs [hunk 17]
```diff
@@ -0,0 +1,93 @@
...
+            }
+            
+            Network = peerNetworkInfo;
+            
+            Protocols = new Dictionary<string, EthProtocolInfo>();
+            Protocols.Add("eth", ethProtocolInfo);
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.JsonRpc/Modules/Parity/PeerInfo.cs:90]: how about other protocols?

File: src/Nethermind/Nethermind.JsonRpc/Modules/Parity/PeerInfo.cs (hunk 15 above)
Comment (tkstanczak) [reply src/Nethermind/Nethermind.JsonRpc/Modules/Parity/PeerInfo.cs:77]: hmmm we need to dynamically detect protocols (wit, eth, pip, les, etc...)

File: src/Nethermind/Nethermind.JsonRpc/Modules/Parity/PeerInfo.cs (hunk 17 above)
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.JsonRpc/Modules/Parity/PeerInfo.cs:90]: parity doesn't support it, we can but didn't for now

**Comment by faderboos** [comment]
👀

File: src/Nethermind/Nethermind.JsonRpc/Modules/Parity/ParityModule.cs (hunk 13 above)
Comment (faderboos) [reply src/Nethermind/Nethermind.JsonRpc/Modules/Parity/ParityModule.cs:45]: 👀
